File: 2021/Day 01/main.py
from sys import argv

with open(argv[1] if len(argv) > 1 else "input") as f:
    lines = [int(x) for x in f.readlines()]

part1 = 0
prev = None
for line in lines:
    if prev != None and line > prev:
        part1 += 1
    prev = line

print(f"Part 1 = {part1}")

part2 = 0
prev = None
windows = [ sum(x) for x in zip(lines, lines[1:], lines[2:]) ]
for window in windows:
    if prev != None and window > prev:
        part2 += 1
    prev = window

print(f"Part 2 = {part2}")


# Compact version:

# with open(argv[1] if len(argv) > 1 else "input") as f: 
#     lines = [int(x) for x in f.readlines()]
#     print(f"Part 1 = {sum([ int(b > a) for (a,b) in zip(lines, lines[1:]) ])}")
#     print(f"Part 2 = {sum([ int(b > a) for (a,b) in (lambda x: zip(x, x[1:]))([ sum(x) for x in zip(lines, lines[1:], lines[2:]) ]) ])}")

File: 2021/Day 01/main.hs
import System.Environment ( getArgs )

getLines :: FilePath -> IO [String]
getLines filename = lines <$> readFile filename

part1 :: [Int] -> Int
part1 lines = foldr (\(a,b) acc -> if b > a then succ acc else acc) 0 (zip lines (tail lines))

part2 :: [Int] -> Int
part2 lines = foldr (\(a,b) acc -> if b > a then succ acc else acc) 0 (zip windows (tail windows))
    where windows = [ a + b + c | (a,b,c) <- zip3 lines (tail lines) (tail (tail lines)) ]

main :: IO ()
main = do
    args <- getArgs
    let filename = if not (null args) then head args else "input"
    lines <- map read <$> getLines filename
    
    let ans1 = part1 lines
    putStrLn $ unwords ["Part 1:", show ans1]

    let ans2 = part2 lines
    putStrLn $ unwords ["Part 1:", show ans2]